crates/polars-core/src/chunked_array/list/mod.rs

```diff
@@ -47,6 +47,25 @@ impl ListChunked {
         unsafe { Series::from_chunks_and_dtype_unchecked(self.name(), chunks, &self.inner_dtype()) }
     }
 
+    /// Returns an iterator over the offsets of this chunked array.
+    ///
+    /// The offsets are returned as though the array consisted of a single chunk.
+    pub fn iter_offsets(&self) -> impl Iterator<Item = i64> + '_ {
+        let mut offsets = self.downcast_iter().map(|arr| arr.offsets());
+
+        let offsets_first_chunk = offsets.next().unwrap();
+
+        // Make sure the offsets in the other chunks continue where the previous chunk left off.
+        let mut prev = *offsets_first_chunk.last();
+        let offsets_rest = offsets.flat_map(move |buf| {
+            let adjust = prev - buf.first();
+            prev = buf.last() + adjust;
+            buf.iter().map(move |v| v + adjust).skip(1)
+        });
+
+        offsets_first_chunk.iter().copied().chain(offsets_rest)
```

Comment:
Chain, skip and friends compile rather expensive iterators as they will contain more branches than needed. I think it is better to make a utlility struct that `impl Iterator`.

Reply:
Indeed the actual type of the iterator is _super_ complex...

I can make an `OffsetsIterator` helper struct, but then how do I instantiate it / what should the inner type be?

Reply:
I've worked it down to this but I can't actually get it to compile :confused: 

```rust

pub struct OffsetsIterator<T, N>
where
    T: Iterator<Item = i64>,
    N: Iterator<Item = T>,
{
    offsets_iters: N,
    current_offsets_iter: T,
    current_adjusted_offset: i64,
    offset_adjustment: i64,
}

impl<T, N> Iterator for OffsetsIterator<T, N>
where
    T: Iterator<Item = i64>,
    N: Iterator<Item = T>,
{
    type Item = i64;

    fn next(&mut self) -> Option<Self::Item> {
        if let Some(offset) = self.current_offsets_iter.next() {
            self.current_adjusted_offset = offset + self.offset_adjustment;
            Some(self.current_adjusted_offset)
        } else {
            self.current_offsets_iter = self.offsets_iters.next()?;
            let first = self.current_offsets_iter.next().unwrap();
            self.offset_adjustment = self.current_adjusted_offset - first;
            self.next()
        }
    }
}

impl<T, N> OffsetsIterator<T, N>
where
    T: Iterator<Item = i64>,
    N: Iterator<Item = T>,
{
    fn new(ca: &ListChunked) -> Self {
        let mut offsets = ca.downcast_iter().map(|arr| arr.offsets().iter().copied());
        let first_iter = offsets.next().unwrap();
        Self {
            current_offsets_iter: first_iter,
            current_adjusted_offset: 0,
            offset_adjustment: 0,
            offsets_iters: offsets,
        }
    }
}

```

Reply:
Will take a look.